impl/src/main/java/org/jboss/weld/manager/BeanManagerImpl.java

```diff
@@ -1559,4 +1569,160 @@ public void validationFailed(Exception failure, Environment environment) {
+                    continue;
+                }
+                // Has a name
+                if (bean.getName() != null) {
+                    continue;
+                }
+                // The type is excluded
+                if (excludeType != null && excludeType.matcher(bean.getBeanClass().getName()).matches()) {
+                    continue;
+                }
+                // Declares an observer
+                if (beanToDeclaredObservers.containsKey(bean)) {
+                    continue;
+                }
+                // Declares a used producer - see also second pass
+                if (beanToDeclaredProducers.containsKey(bean)) {
+                    continue;
+                }
+                // Is resolved for an injection point
+                if (validator.isResolved(bean) || (metadataExtension != null
+                        && (metadataExtension.isInjectedByEEComponent(bean, this) || metadataExtension.isInstanceResolvedBean(bean, this)))) {
+                    continue;
+                }
+                // The type is annotated with an exclude annotation
+                if (excludeAnnotation != null && hasExcludeAnnotation(bean, excludeAnnotation)) {
+                    continue;
+                }
+                // This bean is very likely an unused producer
+                if (bean instanceof AbstractProducerBean) {
+                    unusedProducers.add(bean);
+                }
+                BootstrapLogger.LOG.dropUnusedBeanMetadata(bean);
+                removable.add(bean);
+            }
+
+            // Second pass to find beans which themself are unused and declare only unused producers
+            if (!unusedProducers.isEmpty()) {
+                for (Bean<?> bean : enabledBeans) {
+                    if (bean instanceof ForwardingBean) {
+                        // Unwrap the forwarding implementation
+                        ForwardingBean<?> forwarding = (ForwardingBean<?>) bean;
+                        bean = forwarding.delegate();
+                    }
+                    if (!beanToDeclaredProducers.containsKey(bean)) {
+                        continue;
+                    }
+                    if (!unusedProducers.containsAll(beanToDeclaredProducers.get(bean))) {
+                        continue;
+                    }
+                    BootstrapLogger.LOG.dropUnusedBeanMetadata(bean);
+                    removable.add(bean);
+                }
+            }
+
+            if (!removable.isEmpty()) {
+                // First remove unused beans from BeanManager
+                enabledBeans.removeAll(removable);
+                sharedBeans.removeAll(removable);
+                // Then perform additional cleanup
+                beanResolver.clear();
+                for (Bean<?> bean : removable) {
+                    if (bean instanceof RIBean<?>) {
+                        // This will be skipped in WeldStartup.endInitialization()
```

Comment:
Why is this going to be skipped in `WeldStartup`?

Reply:
Because in `WeldStartup#endInitialization()` we only call `cleanupAfterBoot()` for "present" beans. `BeanManagerImpl#cleanupAfterBoot()` is invoked after bootstrap services cleanup but before `RIBean#cleanupAfterBoot()`.